src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -153,8 +153,17 @@ private function autowireCalls(\ReflectionClass $reflectionClass, bool $isRoot,
         $this->decoratedClass = null;
         $this->getPreviousValue = null;
 
-        if ($isRoot && ($definition = $this->container->getDefinition($this->currentId)) && null !== ($this->decoratedId = $definition->innerServiceId) && $this->container->has($this->decoratedId)) {
-            $this->decoratedClass = $this->container->findDefinition($this->decoratedId)->getClass();
+        if ($isRoot && ($definition = $this->container->getDefinition($this->currentId)) && $definition->getDecoratedService() && null !== ($innerId = $definition->getDecoratedService()[0]) && $this->container->has($innerId)) {
+            // If the class references to itself and is decorated, provide the inner service id and class to not get a circular reference
+            $this->decoratedClass = $this->container->findDefinition($innerId)->getClass();
+            $decoratedDefinition = $this->container->getDefinition($this->currentId);
+            [$_, $innerId] = $decoratedDefinition->getDecoratedService();
+
+            if (null === $innerId) {
+                $innerId = $this->currentId.'.inner';
+            }
+
+            $this->decoratedId = $innerId;
```

Comment:
```suggestion
            $this->decoratedId = $decoratedDefinition[1] ?? $this->currentId.'.inner';
```

and move `$decoratedDefinition` in the if block 
```diff
- if ($isRoot && ($definition = $this->container->getDefinition($this->currentId)) && $definition->getDecoratedService() && null !== ($innerId = $definition->getDecoratedService()[0]) && $this->container->has($innerId)) {
+ if ($isRoot && ($definition = $this->container->getDefinition($this->currentId)) && ($decoratedDefinition = $definition->getDecoratedService()) && null !== ($innerId = $definition->getDecoratedService()[0]) && $this->container->has($innerId)) {

Reply:
Fixed